### DISCUSSION TITLE
Idiomatic way to check equality of `PyTreeDef` instances

### DISCUSSION THREAD

**Discussion by femtomc** [jax-ml/jax#13521] [Q&A] [answered]
Hi all!

I'm working on adding some static checks to a library I'm developing. One such static check requires that I do a comparison between two `PyTreeDef` instances which I get from structures known at tracing time.

Is there an idiomatic way to:

1. Check if two `PyTreeDef` instances are equivalent? Will `==` equality work for different instances?
2. Check if one `PyTreeDef` defines a tree template which is a prefix of another?

Thank you -- sorry if I've missed the answers in the documentation.

**Comment by froystig** [comment, marked as answer]
`==` works on `PyTreeDef` instances. For example:

```python
>>> _, t1 = jax.tree_util.tree_flatten([1, (2, 3), 4])
>>> _, t2 = jax.tree_util.tree_flatten([5, (6, 7), 8])
>>> _, t3 = jax.tree_util.tree_flatten([5, (6, 7), []])
>>> t1 == t2
True
>>> t1 == t3
False
```

**Reply by jakevdp** [reply]
If the elements are arrays rather than scalars, using `==` will lead to an error. For that case, you might want something like this instead:
```python
jax.tree_util.tree_all(jax.tree_map(jnp.allclose, x, y)) 
```

**Reply by froystig** [reply]
That's important to know when checking equality of individual pytrees. My response meant to cover comparing pytree *defs* (e.g. the second output of `tree_flatten`).
